src/gui/widgets/spinner.cpp

```diff
@@ -150,26 +221,60 @@ spinner_definition::resolution::resolution(const config& cfg)
 namespace implementation
 {
 
-builder_spinner::builder_spinner(const config& cfg)
-	: implementation::builder_styled_widget(cfg)
-{
-}
+	builder_spinner::builder_spinner(const config& cfg)
+		: implementation::builder_styled_widget(cfg)
+		  , step_size_(cfg["step_size"].to_int(1))
+	{
+		if(cfg.has_attribute("minimum_value")) {
+			minimum_value_ = cfg["minimum_value"].to_int();
+		}
+		if(cfg.has_attribute("maximum_value")) {
+			maximum_value_ = cfg["maximum_value"].to_int();
+		}
+		VALIDATE(((!minimum_value_) || (!maximum_value_) || (*minimum_value_ <= *maximum_value_)),
+				"minimum_value (" + std::to_string(*minimum_value_) + ") must be <= maximum_value (" + std::to_string(*maximum_value_) + ").");
 
-std::unique_ptr<widget> builder_spinner::build() const
-{
-	auto widget = std::make_unique<spinner>(*this);
+		if(cfg.has_attribute("value")) {
+			value_ = cfg["value"].to_int();
+		} else {
+			if((minimum_value_) && (maximum_value_)) {
+				value_ = ((*minimum_value_ + *maximum_value_) / 2);
```

Comment:
Why not set this 0 in both cases? Since the user can type in any number, it shouldn't matter much.

Reply:
If max and/or min are set, we can't assume 0 to be valid.  And if both are set, starting right in the middle is a decent default if the user doesn't specify.  0 (or the min) is probably going to be the best default in most cases, but in some it would seem odd, while the midpoint is kind of the least-worst option for all cases.  I suspect it really doesn't matter much, since it's so easy for the user to just pick a starting value.

Reply:
> I suspect it really doesn't matter much, since it's so easy for the user to just pick a starting value.

why complicate the code then?

Reply:
I've looked at a couple other widget toolkits, and they all by default start at zero and also allow to specify the initial value in the constructor (equivalent would be a `initial_value` key in WML instance)

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/number
https://docs.oracle.com/javase/8/javafx/api/javafx/scene/control/Spinner.html
https://docs.gtk.org/gtk3/ctor.Adjustment.new.html
https://tgui.eu/tutorials/latest-stable/widgets-overview/#SpinControl

Reply:
I think it makes the most sense to start at 0. If 0 is not within the [min_value, max_value] range, I'd start at min_value if both bounds are positive, otherwise max_value.

Reply:
> > I suspect it really doesn't matter much, since it's so easy for the user to just pick a starting value.
> 
> why complicate the code then?

When the slider represents a change, 0 makes perfect sense for a starting place.  When it represents a range, starting in the middle will often be correct (though this is in a way just another way of starting at 0, if you look at the value as offset from the middle).  But I don't care either way really, 0/min is fine.